File: app/Models/Project.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Carbon\Carbon;

class Project extends Model
{
    use HasFactory;

    protected $fillable = [
        'name',
        'description',
        'status',
        'start_date',
        'end_date',
        'budget',
        'bank_guarantee_amount',
        'bank_guarantee_type',
        'client_name',
        'location',
        'project_manager',
        'project_number',
        'government_entity',
        'consulting_office',
        'project_scope',
        'project_manager_id',
        'progress'
    ];

    protected $casts = [
        'start_date' => 'datetime',
        'end_date' => 'datetime',
        'budget' => 'decimal:2'
    ];

    // Status constants
    const STATUS_PLANNING = 'planning';
    const STATUS_ACTIVE = 'active';
    const STATUS_ON_HOLD = 'on_hold';
    const STATUS_COMPLETED = 'completed';
    const STATUS_CANCELLED = 'cancelled';

    // Get status label in Arabic
    public function getStatusLabelAttribute()
    {
        return match ($this->status) {
            self::STATUS_PLANNING => 'في التخطيط',
            self::STATUS_ACTIVE => 'نشط',
            self::STATUS_ON_HOLD => 'معلق',
            self::STATUS_COMPLETED => 'مكتمل',
            self::STATUS_CANCELLED => 'ملغي',
            default => 'غير محدد'
        };
    }

    // Get status color
    public function getStatusColorAttribute()
    {
        return match ($this->status) {
            self::STATUS_PLANNING => 'blue',
            self::STATUS_ACTIVE => 'green',
            self::STATUS_ON_HOLD => 'yellow',
            self::STATUS_COMPLETED => 'gray',
            self::STATUS_CANCELLED => 'red',
            default => 'gray'
        };
    }

    // Calculate duration in days
    public function getDurationAttribute()
    {
        if ($this->end_date && $this->start_date) {
            return $this->start_date->diffInDays($this->end_date);
        }
        return null;
    }

    // Check if project is active
    public function getIsActiveAttribute()
    {
        return in_array($this->status, [self::STATUS_PLANNING, self::STATUS_ACTIVE]);
    }

    // Format budget with currency
    public function getFormattedBudgetAttribute()
    {
        return number_format((float)$this->budget, 0) . ' ر.س';
    }

    // Format bank guarantee amount with currency
    public function getFormattedBankGuaranteeAmountAttribute()
    {
        if ($this->bank_guarantee_amount) {
            return number_format((float)$this->bank_guarantee_amount, 0) . ' ر.س';
        }
        return 'غير محدد';
    }

    // Get bank guarantee type in Arabic
    public function getBankGuaranteeTypeNameAttribute()
    {
        switch ($this->bank_guarantee_type) {
            case 'cash':
                return 'كاش';
            case 'facilities':
                return 'تسهيلات';
            case 'performance':
                return 'أداء';
            case 'advance_payment':
                return 'سلفة';
            case 'maintenance':
                return 'صيانة';
            case 'other':
                return 'أخرى';
            default:
                return 'غير محدد';
        }
    }

    // Calculate days remaining
    public function getDaysRemainingAttribute()
    {
        if ($this->end_date && $this->status === self::STATUS_ACTIVE) {
            $remaining = Carbon::now()->diffInDays($this->end_date, false);
            return max(0, $remaining);
        }
        return null;
    }

    // Relationships
    public function projectManager()
    {
        return $this->belongsTo(Employee::class, 'project_manager_id');
    }

    public function projectFiles()
    {
        return $this->hasMany(ProjectFile::class);
    }

    public function projectImages()
    {
        return $this->hasMany(ProjectImage::class);
    }

    public function deliveryRequests()
    {
        return $this->hasMany(ProjectDeliveryRequest::class);
    }

    public function projectItems()
    {
        return $this->hasMany(ProjectItem::class);
    }

    public function projectExtracts()
    {
        return $this->hasMany(ProjectExtract::class);
    }

    public function locations()
    {
        return $this->hasMany(Location::class, 'project_id');
    }

    public function equipment()
    {
        return $this->hasManyThrough(Equipment::class, Location::class, 'project_id', 'location_id');
    }

    public function extensions()
    {
        return $this->hasMany(ProjectExtension::class);
    }

    public function visits()
    {
        return $this->hasMany(ProjectVisit::class);
    }

    public function rentalEquipment()
    {
        return $this->hasMany(ProjectRentalEquipment::class);
    }

    public function loans()
    {
        return $this->hasMany(ProjectLoan::class);
    }

    // Scopes
    public function scopeActive($query)
    {
        return $query->where('status', 'active');
    }
}

File: abraj_mobile/lib/models/project.dart
import 'package:flutter/material.dart';

class Project {
  final int id;
  final String name;
  final String? description;
  final DateTime startDate;
  final DateTime? endDate;
  final String status;
  final double? budget;
  final int? locationId;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Project({
    required this.id,
    required this.name,
    this.description,
    required this.startDate,
    this.endDate,
    required this.status,
    this.budget,
    this.locationId,
    this.createdAt,
    this.updatedAt,
  });

  factory Project.fromJson(Map<String, dynamic> json) {
    return Project(
      id: json['id'],
      name: json['name'],
      description: json['description'],
      startDate: DateTime.parse(json['start_date']),
      endDate:
          json['end_date'] != null ? DateTime.parse(json['end_date']) : null,
      status: json['status'],
      budget: json['budget'] != null
          ? double.tryParse(json['budget'].toString())
          : null,
      locationId: json['location_id'],
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'start_date': startDate.toIso8601String().split('T')[0],
      'end_date': endDate?.toIso8601String().split('T')[0],
      'status': status,
      'budget': budget,
      'location_id': locationId,
      'created_at': createdAt?.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }

  String get statusDisplayName {
    switch (status) {
      case 'planning':
        return 'التخطيط';
      case 'active':
        return 'نشط';
      case 'completed':
        return 'مكتمل';
      case 'suspended':
        return 'معلق';
      default:
        return status;
    }
  }

  Color get statusColor {
    switch (status) {
      case 'planning':
        return Colors.orange;
      case 'active':
        return Colors.green;
      case 'completed':
        return Colors.blue;
      case 'suspended':
        return Colors.red;
      default:
        return Colors.grey;
    }
  }
}
